lib/c.c

```diff
@@ -724,3 +724,68 @@ void free(void *ptr)
     __freelist_head->prev = cur;
     __freelist_head = cur;
 }
+
+void snprintf(char *buffer, size_t n, char *str, ...)
+{
+    int *var_args = &str + 4;
+    int si = 0, bi = 0, pi = 0;
+
+    while (str[si]) {
+        if (bi >= n-1) {
+            break;
+        } else if (str[si] != '%') {
+            buffer[bi] = str[si];
+            bi++;
+            si++;
+        } else {
+            int w = 0, zp = 0, pp = 0;
+
+            si++;
+            if (str[si] == '#') {
+                pp = 1;
+                si++;
+            }
+            if (str[si] == '0') {
+                zp = 1;
+                si++;
+            }
+            if (str[si] >= '1' && str[si] <= '9') {
+                w = str[si] - '0';
+                si++;
+                if (str[si] >= '0' && str[si] <= '9') {
+                    w = w * 10;
+                    w += str[si] - '0';
+                    si++;
+                }
+            }
+            switch (str[si]) {
+            case 37: /* % */
+                buffer[bi++] = '%';
+                si++;
+                continue;
+            case 99: /* c */
+                buffer[bi++] = var_args[pi];
+                break;
+            case 115: /* s */
+                strcpy(buffer + bi, var_args[pi]);
+                bi += strlen(var_args[pi]);
+                break;
+            case 111: /* o */
+                bi += __format(buffer + bi, var_args[pi], w, zp, 8, pp);
+                break;
+            case 100: /* d */
+                bi += __format(buffer + bi, var_args[pi], w, zp, 10, 0);
+                break;
+            case 120: /* x */
+                bi += __format(buffer + bi, var_args[pi], w, zp, 16, pp);
+                break;
+            default:
+                abort();
+                break;
+            }
+            pi++;
+            si++;
+        }
+    }
+    if (bi < n) buffer[bi] = '\0';
+}
```

Comment:
When using git for version control, it is recommended to add a newline at the end of files. This is because git compares files line by line, and the end of each line is identified by a newline character.

For more details, refer to the following link:
https://medium.com/@alexey.inkin/how-to-force-newline-at-end-of-files-and-why-you-should-do-it-fdf76d1d090e